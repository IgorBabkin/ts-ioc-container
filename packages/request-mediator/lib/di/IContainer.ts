import { constructor } from 'ts-constructor-injector';

export interface IContainer {
    registerValue(key: string | symbol, value: unknown): void;

    createScope(tags: string[]): IContainer;

    resolve<T>(QueryHandler: constructor<T> | symbol): T;

    dispose(): void;
}
