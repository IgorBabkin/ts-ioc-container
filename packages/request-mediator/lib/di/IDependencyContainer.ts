import { constructor } from '../others';

export interface IDependencyContainer {
  registerValue(key: string | symbol, value: unknown): void;

  createScope(tags: string[]): IDependencyContainer;

  resolve<T>(key: constructor<T> | symbol): T;

  dispose(): void;

  onBeforeDispose(): Promise<void>;
}
