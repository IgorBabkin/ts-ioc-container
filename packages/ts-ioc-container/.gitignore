.idea
cjm
esm
coverage
node_modules
coverage
yarn.log
