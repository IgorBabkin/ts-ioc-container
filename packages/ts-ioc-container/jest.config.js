module.exports = {
    testMatch: ['**/?(*.)+(spec|test).+(ts)'],
    transform: {
        '^.+\\.tsx?$': 'ts-jest',
    },
    collectCoverage: true,
    coveragePathIgnorePatterns: ['/node_modules/', '/fixtures/', '/cjm/', '/esm/'],
    moduleDirectories: ['node_modules', '<rootDir>/lib'],
}
