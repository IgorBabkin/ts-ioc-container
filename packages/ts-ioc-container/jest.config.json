{
  "testMatch": [
    "**/?(*.)+(spec|test).+(ts)"
  ],
  "transform": {
    "^.+\\.tsx?$": [
      "ts-jest",
      {
        "tsConfig": "./tsconfig.json",
        "isolatedModules": false
      }
    ]
  },
  "coveragePathIgnorePatterns": [
    "/node_modules/",
    "cjm/",
    "esm/",
    "scripts/"
  ],
  "moduleDirectories": [
    "node_modules"
  ]
}
