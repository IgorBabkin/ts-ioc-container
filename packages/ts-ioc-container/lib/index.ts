export {
  IContainer,
  Resolvable,
  IContainerModule,
  isDependencyKey,
  DependencyKey,
  InjectionToken,
  Tag,
  Tagged,
} from './container/IContainer';
export { constructor } from './utils';
export { Container } from './container/Container';
export { EmptyContainer } from './container/EmptyContainer';
export { ResolveDependency, IProvider, visible, provider } from './provider/IProvider';
export { IInjector } from './injector/IInjector';
export { DependencyNotFoundError } from './errors/DependencyNotFoundError';
export { MethodNotImplementedError } from './errors/MethodNotImplementedError';
export { ContainerDisposedError } from './errors/ContainerDisposedError';
export { Provider } from './provider/Provider';
export { ProviderDecorator } from './provider/ProviderDecorator';
export { ArgsFn, argsFn, args, ArgsProvider } from './provider/ArgsProvider';
export { singleton, SingletonProvider } from './provider/SingletonProvider';
export { AutoMockedContainer } from './container/AutoMockedContainer';
export { Registration } from './registration/Registration';
export { key, alias, IRegistration, scope, register } from './registration/IRegistration';
export { MetadataInjector, inject } from './injector/MetadataInjector';
export { SimpleInjector } from './injector/SimpleInjector';
export { ProxyInjector } from './injector/ProxyInjector';
export { getHooks, hook, hasHooks } from './hook';
export { by, InstancePredicate } from './by';
export {
  setMetadata,
  getMetadata,
  setParameterMetadata,
  setMethodMetadata,
  getMethodMetadata,
  getParameterMetadata,
} from './metadata';
