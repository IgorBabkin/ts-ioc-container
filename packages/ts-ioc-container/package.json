{
  "name": "ts-ioc-container",
  "version": "8.1.0",
  "description": "Typescript IoC container",
  "author": "ibabkin <ibabkin@goodgamestudios.com>",
  "homepage": "https://github.com/IgorBabkin/service-locator/tree/master/packages/ts-ioc-container",
  "license": "ISC",
  "sideEffects": false,
  "main": "cjm/index.js",
  "types": "esm/index.d.ts",
  "module": "esm/index.js",
  "keywords": [
    "typescript",
    "ioc",
    "service-locator",
    "di"
  ],
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "cjm/**/*",
    "esm/**/*"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/IgorBabkin/service-locator"
  },
  "bugs": {
    "url": "https://github.com/IgorBabkin/service-locator/issues"
  },
  "scripts": {
    "build:cjs": "rimraf cjm && tsc -p tsconfig.json --outDir cjm --module CommonJS --target ES5",
    "build:esm": "rimraf esm && tsc -p tsconfig.json --outDir esm",
    "build": "npm-run-all -p build:**",
    "watch": "nodemon --watch ./lib --exec npm run build",
    "test": "jest"
  },
  "devDependencies": {
    "@types/jest": "27.0.2",
    "jest": "27.2.3",
    "moq.ts": "^7.3.4",
    "nodemon": "^2.0.13",
    "npm-run-all": "4.1.5",
    "reflect-metadata": "^0.1.13",
    "rimraf": "3.0.2",
    "ts-jest": "27.0.5",
    "typescript": "4.4.3"
  },
  "gitHead": "22a23a7b9e7cb9260e1a8858412ddadd6421a53a",
  "peerDependencies": {
    "reflect-metadata": "^0.1.13"
  }
}
