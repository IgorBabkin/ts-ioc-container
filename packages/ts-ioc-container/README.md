# Typescript IoC (Inversion Of Control) container

![NPM version:latest](https://img.shields.io/npm/v/ts-ioc-container/latest.svg?style=flat-square)
![npm downloads](https://img.shields.io/npm/dt/ts-ioc-container.svg?style=flat-square)
![npm bundle size (minified + gzip)](https://img.shields.io/bundlephobia/minzip/ts-ioc-container)
![License](https://img.shields.io/npm/l/ts-ioc-container)

## Advantages
- battle tested :boom:
- written on `typescript`
- simple and lightweight (roughly it's just one file of **~100 lines**) :heart:
- clean API :green_heart:
- supports `tagged scopes`
- fully test covered :100:
- can be used with decorators `@inject`
- composable and open to extend
- awesome for testing (auto mocking)

## Install
```shell script
npm install ts-ioc-container ts-constructor-injector reflect-metadata
```
```shell script
yarn add ts-ioc-container ts-constructor-injector reflect-metadata
```

## tsconfig.json
```json
{
  "compilerOptions": {
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true
  }
}
```


## Injector
How to create new container

```typescript
import { Container, IContainer, IInjector, Provider } from "ts-ioc-container";
import { resolve } from 'ts-constructor-injector';

class Logger {
  info(message: string) {
    console.log(message);
  }
}

const injector: IInjector = {
  resolve<T>(container: IContainer, value: constructor<T>, ...deps: unknown[]): T {
    return resolve(container)(value, ...deps);
  },
};

const container = new Container(injector)
  .register('ILogger', Provider.fromClass(Logger));
const logger = container.resolve<ILogger>('ILogger');
logger.info('Hello world');
```

## Provider

```typescript
import { Provider, singleton, tags } from "ts-ioc-container";

const container = new Container(injector, { tags: ['root'] });
container.register('ILogger', new Provider((container, ...args) => new Logger(...args)));

// Available only in root scope and all his children
container.register('ILogger', Provider.fromClass(Logger).pipe(tags('root')));

// Singleton per root tag and all his children
container.register('ILogger', Provider.fromClass(Logger).pipe(singleton(), tags('root')));

// singleton for scope with tag1 or tag2
container.register('ILogger', Provider.fromClass(Logger).pipe(singleton(), tags('tag1', 'tag2')));

// singleton in every scope
container.register('ILogger', Provider.fromClass(Logger).pipe(args('dev'), singleton()));

// singleton in every scope
container.register('ILogger', Provider.fromClass(Logger).pipe(argsFn((scope) => [scope.resolve('isTestEnv') ? 'dev' : 'prod']), singleton()));

container.register('ILogger', Provider.fromValue(new Logger()));
```

## Registration module (Provider + DependencyKey)

```typescript
import { singleton, tags, forKey, Registration } from "ts-ioc-container";

@forKey('ILogger')
@provider(singleton(), tags('root'))
class Logger {
  info(message: string) {
    console.log(message);
  }
}

const container = new Container(injector, { tags: ['root'] })
  .add(Registration.fromClass(Logger));
const logger = container.resolve<ILogger>('ILogger');
logger.info('Hello world');
```

## Decorators

```typescript
import { singleton, tags, forKey, by, Registration } from "ts-ioc-container";
import { inject } from "ts-constructor-injector";

@forKey('IEngine')
@provider(singleton(), tags('root'))
class Engine {
  constructor(@inject(by('ILogger')) private logger: ILogger) {
  }
}

// OR

const perRoot = provider(singleton(), tags('root'))

@perRoot
@forKey('IEngine')
class Engine {
  constructor(@inject(by('ILogger')) private logger: ILogger) {
  }
}

const container = new Container(injector, { tags: ['root'] })
  .add(Registration.fromClass(Engine));
```

## Hooks

```typescript
import {
  Container,
  IInjector,
  ContainerHook,
  Injector,
  Registration,
} from "ts-ioc-container";
import { MethodReflector } from "ts-constructor-injector";

export const onConstructReflector = new MethodReflector('OnConstructHook');
export const onConstruct = onConstructReflector.createMethodHookDecorator();

export const onDisposeReflector = new MethodReflector('OnDisposeHook');
export const onDispose = onDisposeReflector.createMethodHookDecorator();

@forKey('ILogger')
class Logger {
  @onConstruct
  initialize() {
    console.log('initialized');
  }

  @onDispose
  dispose() {
    console.log('disposed');
  }
}

const injector: IInjector = {
  resolve<T>(container: IContainer, value: constructor<T>, ...deps: unknown[]): T {
    const instance = resolve(container)(value, ...deps);
    onConstructReflector.invokeHooksOf(instance)
    return instance;
  },
}

const container = new Container(injector)
  .add(Registration.fromClass(Logger));
const logger = container.resolve<ILogger>('ILogger'); // initialized
for (const instance of container.getInstances()) {
  onDisposeReflector.invokeHooksOf(instance); // disposed
}
```

## Scopes (child containers)

- tags - you can add tag to scope and root container. And register provider per tag.

```typescript
import { composeDecorators } from "ts-constructor-injector";
import { forKey, provider, Registration, singleton } from "ts-ioc-container";

@forKey('IEngine')
@provider(tags('root'), singleton())
class Logger {
}

@forKey('IEngine')
@provider(tags('home'), singleton())
class Engine {
  constructor(@inject(by('ILogger')) private logger: ILogger) {
  }
}

const container = new Container(injector, { tags: ['root'] })
  .add(Registration.fromClass(Logger))
  .add(Registration.fromClass(Engine));

const scope = container.createScope(['home', 'child']);
const logger = scope.resolve('ILogger');
scope.dispose();
```

## Container Modules

```typescript
import { Registration } from "ts-ioc-container";

class Development implements IContainerModule {
  applyTo(container: IContainer): void {
    container.add(Registration.fromClass(DevLogger));
  }
}

class Production implements IContainerModule {
  applyTo(container: IContainer): void {
    container.add(Registration.fromClass(ProdLogger));
  }
}

const container = new Container(injector, { tags: ['root'] })
  .add(Registration.fromClass(Logger))
  .add(process.env.NODE_ENV === 'production' ? new Production() : new Development());
```

## Mocking / Tests

```typescript
import {
  AutoMockedContainer,
  Container,
  DependencyKey,
} from "ts-ioc-container";
import { Mock } from "moq.ts";

export class MoqContainer extends AutoMockedContainer {
  private mocks = new Map<DependencyKey, IMock<any>>();

  resolve<T>(key: DependencyKey): T {
    return this.resolveMock<T>(key).object();
  }

  dispose(): void {
    this.mocks.clear();
  }

  resolveMock<T>(key: DependencyKey): IMock<T> {
    if (!this.mocks.has(key)) {
      this.mocks.set(key, new Mock());
    }
    return this.mocks.get(key) as IMock<T>;
  }
}

describe('test', () => {
  const mockContainer = new MoqContainer();
  const container = new Container(injector, { parent: mockContainer });

  const engineMock = mockContainer.resolveMock<IEngine>('IEngine');
  engineMock.setup(i => i.getRegistrationNumber()).return('123');

  const engine = container.resolve<IEngine>('IEngine');

  expect(engine.getRegistrationNumber()).toBe('123');
})
```


## Errors

- [DependencyNotFoundError.ts](lib%2Fcontainer%2FDependencyNotFoundError.ts)
- [DependencyMissingKeyError.ts](lib%2Fregistration%2FDependencyMissingKeyError.ts)
- [MethodNotImplementedError.ts](lib%2FMethodNotImplementedError.ts)
- [ContainerDisposedError.ts](lib%2Fcontainer%2FContainerDisposedError.ts)

