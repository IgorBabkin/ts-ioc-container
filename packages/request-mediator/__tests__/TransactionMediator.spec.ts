import 'reflect-metadata';
import { by, inject, key, provider, register, Registration as R, scope, singleton } from 'ts-ioc-container';
import {
  IQueryHandler,
  ITransactionContext,
  ITransactionContextKey,
  request,
  RequestMediator,
  Scope,
  transaction,
} from '../lib';
import { ContainerAdapter, createContainer, EmptyType } from './di';
import { ILoggerKey, Logger } from './Logger';

let id = 0;

@register(key('Repo'), scope((s) => s.hasTag(Scope.Request)))
@provider(singleton())
class Repo {
  private entities: string[] = [];
  private id = id;
  constructor(
    @inject((s) => () => s.resolve(ITransactionContextKey)) private getTransaction: () => TransactionContext,
    @inject(by.key(ILoggerKey)) private logger: Logger,
  ) {
    id++;
  }

  async findById(): Promise<void> {
    const transaction = this.getTransaction();
    this.entities.push('todo');
    this.logger.addLog(
      `Repo (${this.id}) -> findById [transactionId: ${transaction.id.toString()}], entities: ${this.entities.join(', ')}`,
    );
  }

  async persist(): Promise<void> {
    const transaction = this.getTransaction();
    this.logger.addLog(
      `Repo (${this.id}) -> persist [transactionId: ${transaction.id.toString()}], entities: ${this.entities.join(', ')}`,
    );
  }
}

class TransactionContext implements ITransactionContext {
  constructor(public id = 0) {}

  execute<Response>(fn: (context: ITransactionContext) => Promise<Response>): Promise<Response> {
    return fn(new TransactionContext(this.id + 1));
  }
}

@transaction
class AfterHandler implements IQueryHandler<EmptyType, void> {
  constructor(
    @inject(by.key(ILoggerKey)) private logger: Logger,
    @inject(by.key('Repo')) private repo: Repo,
    @inject(by.key(ITransactionContextKey)) private transactionContext: TransactionContext,
  ) {}

  async handle(query: EmptyType): Promise<void> {
    this.logger.addLog(`AfterHandler - transactionId: ${this.transactionContext.id.toString()}`);
    await this.repo.persist();
  }
}

class BeforeHandler implements IQueryHandler<EmptyType, void> {
  constructor(
    @inject(by.key(ILoggerKey)) private logger: Logger,
    @inject(by.key(ITransactionContextKey)) private transactionContext: TransactionContext,
  ) {}

  async handle(): Promise<void> {
    this.logger.addLog(`BeforeHandler - transactionId: ${this.transactionContext.id.toString()}`);
  }
}

@request('before', [BeforeHandler])
@request('after', [AfterHandler])
class QueryHandler implements IQueryHandler<EmptyType, void> {
  constructor(
    @inject(by.key(ILoggerKey)) private logger: Logger,
    @inject(by.key('Repo')) private repo: Repo,
    @inject(by.key(ITransactionContextKey)) private transactionContext: TransactionContext,
  ) {}

  async handle(query: EmptyType): Promise<void> {
    this.logger.addLog(`QueryHandler - transactionId: ${this.transactionContext.id.toString()}`);
    await this.repo.findById();
  }
}

@register(key(ITransactionContextKey), scope((s) => s.hasTag(Scope.Application)))
@provider(singleton())
class TestTransaction implements ITransactionContext {
  constructor(private id: number = 0) {}

  execute<Response>(fn: (context: ITransactionContext) => Promise<Response>): Promise<Response> {
    return fn(new TestTransaction(this.id + 1));
  }
}

describe('TransactionMediator', () => {
  it('should invoke middleware', async () => {
    const container = createContainer()
      .add(R.fromClass(Logger))
      .add(R.fromClass(TestTransaction))
      .add(R.fromClass(Repo));

    const mediator = new RequestMediator(new ContainerAdapter(container));

    await mediator.send(QueryHandler, {});

    expect(container.resolve<Logger>(ILoggerKey).getLogs()).toMatchSnapshot();
  });
});
