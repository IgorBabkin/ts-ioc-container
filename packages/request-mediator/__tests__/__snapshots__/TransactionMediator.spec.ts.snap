// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionMediator should invoke middleware 1`] = `
Array [
  "BeforeHandler - transactionId: 0",
  "QueryHandler - transactionId: 0",
  "Repo (0) -> findById [transactionId: 0], entities: todo",
  "AfterHandler - transactionId: 1",
  "Repo (0) -> persist [transactionId: 1], entities: todo",
]
`;
